### PR TITLE
feat(registries): add `list-registries` command

### LINKED ISSUE #5001: registries: implement `list-registries`
### What needs to get done

* Create a `list-registry-sets` command
* Create a base `AssertionService`
* Create a child `RegistrySetService`
* Add store calls to `snapcraft.store.client`
* Add spread tests
* Implement command:

```
Usage:
    snapcraft list-registry-sets [options]

Options:
              ...
           --name:  Name of registry set to list
         --format:  The output format (table | json)
```

This is prototyped in https://github.com/canonical/snapcraft/issues/4917

### Why it needs to get done

To allow users to view registry sets

Blocked by #4911

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

Add support for the `list-registries` command.

Fixes #5001 
(CRAFT-3295)

### CODE REVIEW

File: snapcraft/commands/registries.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                const.OutputFormat.table.value,
+                const.OutputFormat.json.value,
+            ],
+            help="The output format (table | json)",
+            default="table",
+        )
+
+    @override
+    def run(self, parsed_args: "argparse.Namespace"):
+        self._services.registries.list_assertions(  # type: ignore[attr-defined,reportAttributeAccessIssue]
```
Comment (mr-cal) [inline snapcraft/commands/registries.py:61]: I cannot figure out how to fix the typing:
```
lint-pyright: commands[0]> pyright
/home/developer/dev/snapcraft/snapcraft/commands/registries.py
  /home/developer/dev/snapcraft/snapcraft/commands/registries.py:61:35 - error: Cannot access attribute "list_assertions" for class "AppService"
    Attribute "list_assertions" is unknown (reportAttributeAccessIssue)
```

Comment (mr-cal) [reply snapcraft/commands/registries.py]: @lengau figured it out!

File: snapcraft/commands/registries.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+from snapcraft import const, services
+
+
+class StoreListRegistriesCommand(craft_application.commands.AppCommand):
+    """List registries."""
+
+    name = "list-registries"
+    help_msg = "List registries"
+    overview = textwrap.dedent(
```
Comment (lengau) [inline snapcraft/commands/registries.py:33]: I have no idea what should be here, but I feel like whatever is here should be more detailed than this. Some ideas:

- Why you would need to know what registries you have
- A link to external documentation about registries.

**Review by lengau — APPROVED** [review]
LGTM!

File: snapcraft/cli.py [hunk 3]
```diff
@@ -132,12 +132,13 @@
         ],
     ),
     craft_cli.CommandGroup(
-        "Store Validation Sets",
```
Comment (sergiusens) [inline snapcraft/cli.py:135]: we should probably keep this name and create a new group for registries.

Assertions is not a word we should expose

File: snapcraft/commands/registries.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/commands/registries.py:33]: Take some inspiration from the spec (SD133)

File: snapcraft/commands/registries.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    _services: services.SnapcraftServiceFactory  # type: ignore[reportIncompatibleVariableOverride]
+
+    @override
+    def fill_parser(self, parser: "argparse.ArgumentParser") -> None:
+        parser.add_argument(
+            "--name",
+            metavar="name",
+            required=False,
+            type=str,
+            help="Name of registry set to list",
```
Comment (sergiusens) [inline snapcraft/commands/registries.py:47]:
```suggestion
            help="Name of the registry set to list",
```

File: snapcraft/commands/registries.py (hunk 2 above)
Comment (mr-cal) [reply snapcraft/commands/registries.py:33]: I asked the snapd team about this last week - they are planning to write documentation this cycle about registries, so eventually I will be able to link to their docs, similar to `list-validation-sets`.

I added a more information.  We should probably improve the other 7 `list-` command overviews.  Only `list-tracks` has more than a few words.

**Review by sergiusens — APPROVED** [review]

**Pull request merged**